libmamba/src/core/download_impl.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_CORE_DOWNLOAD_IMPL_HPP
+#define MAMBA_CORE_DOWNLOAD_IMPL_HPP
+
+#include <chrono>
+#include <unordered_map>
+
+#include "mamba/core/download.hpp"
+
+#include "compression.hpp"
+#include "curl.hpp"
+
+namespace mamba
+{
+    enum class DownloadState
+    {
+        WAITING,
+        PREPARING,
+        RUNNING,
+        FINISHED,
+        FAILED
+    };
+
+    class DownloadAttempt
+    {
+
+    public:
+
+        using completion_function = std::function<bool(CURLMultiHandle&,CURLcode)>;
+        using on_success_callback = std::function<bool(DownloadSuccess)>;
+        using on_failure_callback = std::function<bool(DownloadError)>;
+
+        DownloadAttempt(CURLHandle& handle, const DownloadRequest* request);
+
+        void prepare_download(CURLMultiHandle& downloader,
+                              const Context& context,
+                              on_success_callback success,
+                              on_failure_callback error);
+
+        auto create_completion_function() -> completion_function;
+
+    private:
+
+        bool finish_download(CURLMultiHandle& downloader, CURLcode code);
+        void clean_attempt(CURLMultiHandle& downloader, bool erase_downloaded);
+        void invoke_progress_callback(const DownloadEvent&) const;
+
+        void configure_handle(const Context& context);
+        void configure_handle_headers(const Context& context);
+
+        size_t write_data(char* buffer, size_t data);
+
+        static size_t curl_header_callback(char* buffer, size_t size, size_t nbitems, void* self);
+        static size_t curl_write_callback(char* buffer, size_t size, size_t nbitems, void* self);
+        static int curl_progress_callback(void* f, curl_off_t total_to_download, curl_off_t now_downloaded, curl_off_t, curl_off_t);
+
+        TransferData get_transfer_data() const;
+        DownloadError build_download_error(CURLcode code) const;
+        DownloadError build_download_error(TransferData data) const;
+        DownloadSuccess build_download_success(TransferData data) const;
+
+        CURLHandle* p_handle;
+        const DownloadRequest* p_request;
+        on_success_callback m_success_callback;
+        on_failure_callback m_failure_callback;
+        std::size_t m_retry_wait_seconds;
+        std::unique_ptr<CompressionStream> p_stream;
+        std::ofstream m_file;
+        std::string m_cache_control;
+        std::string m_etag;
+        std::string m_last_modified;
+    };
+
+    class DownloadTracker
+    {
+    public:
+
+        using completion_function = DownloadAttempt::completion_function;
+        using completion_map_entry = std::pair<CURLId, completion_function>;
+        
+        DownloadTracker(const DownloadRequest* request, std::size_t max_retries, bool fail_fast);
```

Comment:
same remark about request should be a const ref here

Reply:
Also to avoid the bool in the interface of DownloadTRacker, consider grouping both value arguments in a DownloadTrackerOptions struct with them and default values.

Reply:
Actually I did it in the first version but I found it a bit overkilled for only two members. I will reintroduce it, I agree the boolean here is kind of ugly.